File: mobile/ios/ShiftPlanner/ShiftPlanner/Shared/Repositories/APIScheduleRepository.swift
import Foundation

final class APIScheduleRepository: ScheduleRepository {
    private let apiClient: APIClient

    init(apiClient: APIClient = .shared) {
        self.apiClient = apiClient
    }

    // MARK: - Generate / fetch

    func generateSchedule(startDate: Date, endDate: Date, branchId: Int?) async throws -> [AppSchedule] {
        let payload = ScheduleGenerateRequestDTO(
            startDate: Self.dateFormatter.string(from: startDate),
            endDate: Self.dateFormatter.string(from: endDate),
            branchId: branchId
        )
        let body = try JSONEncoder().encode(payload)
        let request = apiClient.makeRequest(
            path: "schedule/generate",
            method: "POST",
            body: body,
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleGenerateResponseDTO.self)
        return try response.schedules.map(mapSchedule)
    }

    func fetchSchedule(scheduleId: Int) async throws -> AppSchedule {
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleResponseDTO.self)
        return try mapSchedule(response)
    }

    func fetchLatestSchedule(status: AppScheduleStatus?) async throws -> AppSchedule? {
        var queryItems: [URLQueryItem] = []
        if let status {
            queryItems.append(URLQueryItem(name: "status", value: status.rawValue))
        }
        let request = authorizedGetRequest(path: "schedule/latest", queryItems: queryItems)

        let (data, response): (Data, URLResponse)
        do {
            (data, response) = try await URLSession.shared.data(for: request)
        } catch {
            throw APIClientError.requestFailed(message: error.localizedDescription)
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw APIClientError.invalidResponse
        }

        if httpResponse.statusCode == 404 {
            return nil
        }

        guard (200...299).contains(httpResponse.statusCode) else {
            if let errorResponse = try? JSONDecoder().decode(APIErrorResponse.self, from: data),
               let detail = errorResponse.detail,
               !detail.isEmpty {
                throw APIClientError.requestFailed(message: detail)
            }
            throw APIClientError.requestFailed(
                message: LanguageManager.localizedFormat(
                    "Request failed with status code %d.",
                    "Запрос завершился с кодом %d.",
                    httpResponse.statusCode
                )
            )
        }

        let responseDTO: ScheduleResponseDTO
        do {
            responseDTO = try JSONDecoder().decode(ScheduleResponseDTO.self, from: data)
        } catch {
            throw APIClientError.decodingFailed
        }

        return try mapSchedule(responseDTO)
    }

    func fetchSchedules(
        startDate: Date?,
        endDate: Date?,
        branchId: Int?,
        status: AppScheduleStatus?
    ) async throws -> [AppScheduleListItem] {
        var queryItems: [URLQueryItem] = []
        if let branchId {
            queryItems.append(URLQueryItem(name: "branch_id", value: String(branchId)))
        }
        if let startDate {
            queryItems.append(URLQueryItem(name: "start_date", value: Self.dateFormatter.string(from: startDate)))
        }
        if let endDate {
            queryItems.append(URLQueryItem(name: "end_date", value: Self.dateFormatter.string(from: endDate)))
        }
        if let status {
            queryItems.append(URLQueryItem(name: "status", value: status.rawValue))
        }
        let request = authorizedGetRequest(path: "schedule", queryItems: queryItems)
        let response = try await apiClient.send(request, as: [ScheduleListItemResponseDTO].self)
        return response.compactMap(mapListItem)
    }

    func publishSchedule(scheduleId: Int) async throws -> AppSchedule {
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)/publish",
            method: "POST",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleResponseDTO.self)
        return try mapSchedule(response)
    }

    func fetchMySchedule() async throws -> [AppScheduledShift] {
        // calendar-summary aggregates shifts across ALL published schedules;
        // /schedule/my only sees the latest one (older weeks' shifts vanish).
        // Fall back to /schedule/my for older backend deployments.
        if let summary = try? await fetchMyCalendarSummary() {
            return summary.shifts.compactMap { shift in
                guard let date = Self.dateFormatter.date(from: shift.date) else { return nil }
                return AppScheduledShift(
                    id: shift.shiftId,
                    employeeId: summary.employee.id,
                    employeeName: summary.employee.fullName,
                    positionId: summary.employee.position?.id ?? 0,
                    positionName: summary.employee.position?.name ?? "",
                    date: date,
                    startMinutes: Self.minutes(from: shift.startTime),
                    endMinutes: Self.minutes(from: shift.endTime)
                )
            }
        }

        let request = apiClient.makeRequest(
            path: "schedule/my",
            method: "GET",
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: [ScheduleShiftResponseDTO].self)
        return try response.map(mapShift)
    }

    private func fetchMyCalendarSummary() async throws -> EmployeeCalendarSummaryDTO {
        let request = apiClient.makeRequest(
            path: "employees/me/calendar-summary",
            method: "GET",
            requiresAuthorization: true
        )
        return try await apiClient.send(request, as: EmployeeCalendarSummaryDTO.self)
    }

    // MARK: - Assignment

    func fetchAvailableEmployees(
        scheduleId: Int,
        shift: AppScheduledShift,
        branchId: Int?,
        includeUnavailable: Bool
    ) async throws -> [AppAvailableEmployee] {
        var queryItems: [URLQueryItem] = [
            URLQueryItem(name: "date", value: Self.dateFormatter.string(from: shift.date)),
            URLQueryItem(name: "start_time", value: Self.timeString(fromMinutes: shift.startMinutes)),
            URLQueryItem(name: "end_time", value: Self.timeString(fromMinutes: shift.endMinutes)),
            URLQueryItem(name: "position_id", value: String(shift.positionId)),
            URLQueryItem(name: "include_unavailable", value: includeUnavailable ? "true" : "false")
        ]
        if let branchId {
            queryItems.append(URLQueryItem(name: "branch_id", value: String(branchId)))
        }
        let request = authorizedGetRequest(
            path: "schedule/\(scheduleId)/employees/available",
            queryItems: queryItems
        )
        let response = try await apiClient.send(request, as: [AvailableEmployeeResponseDTO].self)
        return response.map(mapAvailableEmployee)
    }

    func assignRequirement(
        scheduleId: Int,
        requirementId: Int,
        employeeId: Int
    ) async throws -> AppSchedule {
        let body = try JSONEncoder().encode(RequirementAssignRequestDTO(employeeId: employeeId))
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)/requirements/\(requirementId)/assign",
            method: "POST",
            body: body,
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleResponseDTO.self)
        return try mapSchedule(response)
    }

    // MARK: - Shift CRUD

    func createShift(scheduleId: Int, mutation: ScheduleShiftMutation) async throws -> AppSchedule {
        let payload = ManualShiftCreateRequestDTO(
            date: Self.dateFormatter.string(from: mutation.date),
            startTime: Self.timeString(fromMinutes: mutation.startMinutes),
            endTime: Self.timeString(fromMinutes: mutation.endMinutes),
            positionId: mutation.positionId,
            employeeId: mutation.employeeId.flatMap { $0 > 0 ? $0 : nil }
        )
        let body = try JSONEncoder().encode(payload)
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)/shifts",
            method: "POST",
            body: body,
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleResponseDTO.self)
        return try mapSchedule(response)
    }

    func updateShift(
        scheduleId: Int,
        shiftId: Int,
        mutation: ScheduleShiftMutation
    ) async throws -> AppSchedule {
        let resolvedEmployeeId = mutation.employeeId.flatMap { $0 > 0 ? $0 : nil }
        // "reassign" replaces the shift's assignment with employee_id (a nil field
        // means "unassign"). Never send "remove" here: on the backend it deletes
        // the shift entirely, which used to make editing an unassigned shift
        // silently delete it.
        let payload = ScheduleShiftUpdateRequestDTO(
            date: Self.dateFormatter.string(from: mutation.date),
            startTime: Self.timeString(fromMinutes: mutation.startMinutes),
            endTime: Self.timeString(fromMinutes: mutation.endMinutes),
            positionId: mutation.positionId,
            employeeId: resolvedEmployeeId,
            action: "reassign"
        )
        let body = try JSONEncoder().encode(payload)
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)/shifts/\(shiftId)",
            method: "PATCH",
            body: body,
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleResponseDTO.self)
        return try mapSchedule(response)
    }

    func deleteSchedule(scheduleId: Int) async throws {
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)",
            method: "DELETE",
            requiresAuthorization: true
        )
        try await apiClient.sendWithoutResponseBody(request)
    }

    func deleteShift(scheduleId: Int, shiftId: Int) async throws -> AppSchedule {
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)/shifts/\(shiftId)",
            method: "DELETE",
            requiresAuthorization: true
        )
        try await apiClient.sendWithoutResponseBody(request)
        return try await fetchSchedule(scheduleId: scheduleId)
    }

    func updateScheduleRequirement(
        scheduleId: Int,
        requirementId: Int,
        mutation: ScheduleShiftMutation,
        quantity: Int
    ) async throws -> AppSchedule {
        let payload = ScheduleRequirementInScheduleUpdateDTO(
            branchId: nil,
            positionId: mutation.positionId,
            date: Self.dateFormatter.string(from: mutation.date),
            minStaff: quantity,
            requiredCount: quantity,
            startTime: Self.timeString(fromMinutes: mutation.startMinutes),
            endTime: Self.timeString(fromMinutes: mutation.endMinutes)
        )
        let body = try JSONEncoder().encode(payload)
        let request = apiClient.makeRequest(
            path: "schedule/\(scheduleId)/requirements/\(requirementId)",
            method: "PATCH",
            body: body,
            requiresAuthorization: true
        )
        let response = try await apiClient.send(request, as: ScheduleResponseDTO.self)
        return try mapSchedule(response)
    }

    // MARK: - Request helper

    private func authorizedGetRequest(path: String, queryItems: [URLQueryItem]) -> URLRequest {
        let baseURL = apiClient.baseURL.appendingPathComponent(path)
        var components = URLComponents(url: baseURL, resolvingAgainstBaseURL: false)
        if !queryItems.isEmpty {
            components?.queryItems = queryItems
        }
        let url = components?.url ?? baseURL
        var request = URLRequest(url: url)
        request.httpMethod = "GET"
        request.setValue("application/json", forHTTPHeaderField: "Content-Type")
        if let accessToken = apiClient.accessToken {
            request.setValue("Bearer \(accessToken)", forHTTPHeaderField: "Authorization")
        }
        return request
    }

    // MARK: - Mapping

    private func mapSchedule(_ dto: ScheduleResponseDTO) throws -> AppSchedule {
        guard let status = AppScheduleStatus(rawValue: dto.status) else {
            throw APIClientError.decodingFailed
        }
        return AppSchedule(
            id: dto.id,
            branchId: dto.branchId,
            startDate: dto.startDate.flatMap { Self.dateFormatter.date(from: $0) },
            endDate: dto.endDate.flatMap { Self.dateFormatter.date(from: $0) },
            status: status,
            shifts: try dto.shifts.map(mapShift),
            unfilledRequirements: try dto.unfilledRequirements.map(mapUnfilledRequirement)
        )
    }

    private func mapListItem(_ dto: ScheduleListItemResponseDTO) -> AppScheduleListItem? {
        guard
            let status = AppScheduleStatus(rawValue: dto.status),
            let start = Self.dateFormatter.date(from: dto.startDate),
            let end = Self.dateFormatter.date(from: dto.endDate)
        else {
            return nil
        }
        return AppScheduleListItem(
            id: dto.id,
            branchId: dto.branchId,
            startDate: start,
            endDate: end,
            status: status
        )
    }

    private func mapShift(_ dto: ScheduleShiftResponseDTO) throws -> AppScheduledShift {
        guard let date = Self.dateFormatter.date(from: dto.date) else {
            throw APIClientError.decodingFailed
        }
        return AppScheduledShift(
            id: dto.id,
            employeeId: dto.employeeId,
            employeeName: dto.employeeName,
            positionId: dto.positionId,
            positionName: dto.position,
            date: date,
            startMinutes: Self.minutes(from: dto.startTime),
            endMinutes: Self.minutes(from: dto.endTime)
        )
    }

    private func mapUnfilledRequirement(_ dto: ScheduleUnfilledRequirementResponseDTO) throws -> AppUnfilledRequirement {
        guard let date = Self.dateFormatter.date(from: dto.date) else {
            throw APIClientError.decodingFailed
        }
        return AppUnfilledRequirement(
            id: dto.requirementId,
            positionId: dto.positionId,
            positionTitle: dto.positionTitle,
            date: date,
            startMinutes: Self.minutes(from: dto.startTime),
            endMinutes: Self.minutes(from: dto.endTime),
            missingStaff: dto.missingStaff
        )
    }

    private func mapAvailableEmployee(_ dto: AvailableEmployeeResponseDTO) -> AppAvailableEmployee {
        AppAvailableEmployee(
            id: dto.id,
            fullName: dto.fullName,
            positionName: dto.position.name,
            branchId: dto.branch?.id,
            branchName: dto.branch?.name,
            availabilityStatus: AppEmployeeAvailabilityStatus.fromApiValue(dto.availabilityStatus),
            assignedHours: dto.assignedHours
        )
    }

    // MARK: - Formatting helpers

    private static func minutes(from timeString: String) -> Int {
        let parts = timeString.split(separator: ":").compactMap { Int($0) }
        guard parts.count >= 2 else { return 0 }
        return (parts[0] * 60) + parts[1]
    }

    static func timeString(fromMinutes minutes: Int) -> String {
        String(format: "%02d:%02d:00", minutes / 60, minutes % 60)
    }

    private static let dateFormatter: DateFormatter = {
        let formatter = DateFormatter()
        formatter.calendar = Calendar(identifier: .gregorian)
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.timeZone = .autoupdatingCurrent
        formatter.dateFormat = "yyyy-MM-dd"
        return formatter
    }()
}

File: mobile/android/app/src/main/java/com/froggyriia/shiftplanner/data/schedule/ApiScheduleRepository.kt
package com.froggyriia.shiftplanner.data.schedule

import com.froggyriia.shiftplanner.data.network.ApiClient
import com.froggyriia.shiftplanner.data.network.AvailableEmployeeResponseDto
import com.froggyriia.shiftplanner.data.network.ManualShiftCreateRequestDto
import com.froggyriia.shiftplanner.data.network.RequirementAssignRequestDto
import com.froggyriia.shiftplanner.data.network.ScheduleGenerateRequestDto
import com.froggyriia.shiftplanner.data.network.ScheduleRequirementInScheduleUpdateDto
import com.froggyriia.shiftplanner.data.network.ScheduleResponseDto
import com.froggyriia.shiftplanner.data.network.ScheduleShiftResponseDto
import com.froggyriia.shiftplanner.data.network.ScheduleShiftUpdateRequestDto
import com.froggyriia.shiftplanner.data.network.ScheduleUnfilledRequirementResponseDto
import com.froggyriia.shiftplanner.domain.model.AppAvailableEmployee
import com.froggyriia.shiftplanner.domain.model.AppEmployeeAvailabilityStatus
import com.froggyriia.shiftplanner.domain.model.AppSchedule
import com.froggyriia.shiftplanner.domain.model.AppScheduleStatus
import com.froggyriia.shiftplanner.domain.model.AppScheduledShift
import com.froggyriia.shiftplanner.domain.model.AppUnfilledRequirement
import com.froggyriia.shiftplanner.domain.model.ScheduleShiftMutation
import java.text.SimpleDateFormat
import java.util.Date
import java.util.Locale

class ApiScheduleRepository(
    private val apiClient: ApiClient
) : ScheduleRepository {

    override suspend fun generateSchedule(startDate: String, endDate: String): List<AppSchedule> = wrap {
        apiClient.api.generateSchedule(
            ScheduleGenerateRequestDto(startDate = startDate, endDate = endDate)
        ).map { it.toDomain() }
    }

    override suspend fun fetchSchedule(scheduleId: Int): AppSchedule = wrap {
        apiClient.api.getSchedule(scheduleId).toDomain()
    }

    override suspend fun fetchLatestSchedule(status: String?): AppSchedule? = wrap {
        val response = apiClient.api.getLatestSchedule(status)
        if (response.code() == 404) return@wrap null
        if (!response.isSuccessful) throw Exception("Request failed with status ${response.code()}")
        response.body()?.toDomain()
    }

    override suspend fun publishSchedule(scheduleId: Int): AppSchedule = wrap {
        apiClient.api.publishSchedule(scheduleId).toDomain()
    }

    override suspend fun fetchMySchedule(): List<AppScheduledShift> = wrap {
        apiClient.api.getMySchedule().map { it.toShift() }
    }

    override suspend fun fetchAvailableEmployees(
        scheduleId: Int,
        shift: AppScheduledShift,
        branchId: Int?,
        includeUnavailable: Boolean
    ): List<AppAvailableEmployee> = wrap {
        apiClient.api.getAvailableEmployees(
            scheduleId = scheduleId,
            date = dateFormatter.format(shift.date),
            startTime = minutesToTime(shift.startMinutes),
            endTime = minutesToTime(shift.endMinutes),
            positionId = shift.positionId,
            branchId = branchId,
            includeUnavailable = includeUnavailable
        ).map { it.toEmployee() }
    }

    override suspend fun assignRequirement(
        scheduleId: Int,
        requirementId: Int,
        employeeId: Int
    ): AppSchedule = wrap {
        apiClient.api.assignRequirement(
            scheduleId = scheduleId,
            requirementId = requirementId,
            request = RequirementAssignRequestDto(employeeId = employeeId)
        ).toDomain()
    }

    override suspend fun createShift(
        scheduleId: Int,
        mutation: ScheduleShiftMutation
    ): AppSchedule = wrap {
        apiClient.api.createShift(
            scheduleId = scheduleId,
            request = ManualShiftCreateRequestDto(
                date = dateFormatter.format(mutation.date),
                startTime = minutesToTime(mutation.startMinutes),
                endTime = minutesToTime(mutation.endMinutes),
                positionId = mutation.positionId,
                employeeId = mutation.employeeId?.takeIf { it > 0 }
            )
        ).toDomain()
    }

    override suspend fun updateShift(
        scheduleId: Int,
        shiftId: Int,
        mutation: ScheduleShiftMutation
    ): AppSchedule = wrap {
        val resolvedEmployeeId = mutation.employeeId?.takeIf { it > 0 }
        val action = when {
            resolvedEmployeeId != null -> "reassign"
            else -> "remove"
        }
        apiClient.api.updateShift(
            scheduleId = scheduleId,
            shiftId = shiftId,
            request = ScheduleShiftUpdateRequestDto(
                date = dateFormatter.format(mutation.date),
                startTime = minutesToTime(mutation.startMinutes),
                endTime = minutesToTime(mutation.endMinutes),
                positionId = mutation.positionId,
                employeeId = resolvedEmployeeId,
                action = action
            )
        ).toDomain()
    }

    override suspend fun deleteShift(scheduleId: Int, shiftId: Int): AppSchedule = wrap {
        apiClient.api.deleteShift(scheduleId = scheduleId, shiftId = shiftId)
        fetchSchedule(scheduleId)
    }

    override suspend fun updateScheduleRequirement(
        scheduleId: Int,
        requirementId: Int,
        mutation: ScheduleShiftMutation,
        quantity: Int
    ): AppSchedule = wrap {
        apiClient.api.updateScheduleRequirement(
            scheduleId = scheduleId,
            requirementId = requirementId,
            request = ScheduleRequirementInScheduleUpdateDto(
                branchId = null,
                positionId = mutation.positionId,
                date = dateFormatter.format(mutation.date),
                minStaff = quantity,
                requiredCount = quantity,
                startTime = minutesToTime(mutation.startMinutes),
                endTime = minutesToTime(mutation.endMinutes)
            )
        ).toDomain()
    }

    // ── Mapping ───────────────────────────────────────────────────────────────

    private fun ScheduleResponseDto.toDomain(): AppSchedule {
        val scheduleStatus = AppScheduleStatus.fromApiValue(status)
            ?: throw Exception("Unknown schedule status: $status")
        return AppSchedule(
            id = id,
            branchId = branchId,
            status = scheduleStatus,
            shifts = shifts.map { it.toShift() },
            unfilledRequirements = unfilledRequirements.map { it.toUnfilled() }
        )
    }

    private fun ScheduleShiftResponseDto.toShift(): AppScheduledShift {
        val parsedDate = dateFormatter.parse(date) ?: Date()
        return AppScheduledShift(
            id = id,
            employeeId = employeeId,
            employeeName = employeeName,
            positionId = positionId,
            positionName = position,
            date = parsedDate,
            startMinutes = timeToMinutes(startTime),
            endMinutes = timeToMinutes(endTime)
        )
    }

    private fun ScheduleUnfilledRequirementResponseDto.toUnfilled(): AppUnfilledRequirement {
        val parsedDate = dateFormatter.parse(date) ?: Date()
        return AppUnfilledRequirement(
            id = requirementId,
            positionId = positionId,
            positionTitle = positionTitle,
            date = parsedDate,
            startMinutes = timeToMinutes(startTime),
            endMinutes = timeToMinutes(endTime),
            missingStaff = missingStaff
        )
    }

    private fun AvailableEmployeeResponseDto.toEmployee(): AppAvailableEmployee {
        val status = AppEmployeeAvailabilityStatus.fromApiValue(availabilityStatus)
        return AppAvailableEmployee(
            id = id,
            fullName = fullName,
            positionName = position.name,
            branchName = branch?.name,
            availabilityStatus = status,
            assignedHours = assignedHours
        )
    }

    private fun timeToMinutes(time: String): Int {
        val parts = time.split(":").mapNotNull { it.toIntOrNull() }
        if (parts.size < 2) return 0
        return parts[0] * 60 + parts[1]
    }

    private fun minutesToTime(minutes: Int): String =
        String.format(Locale.US, "%02d:%02d:00", minutes / 60, minutes % 60)

    private val dateFormatter = SimpleDateFormat("yyyy-MM-dd", Locale.US)

    private suspend fun <T> wrap(block: suspend () -> T): T {
        return try {
            block()
        } catch (error: Throwable) {
            throw Exception(apiClient.userMessage(error))
        }
    }
}
